### PR TITLE
feat: error index reporting implementation

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Queuing **error metadata for failed message IDs** in `err_idx` jobsDB.
- Created **configSubscriber** to **manage config changes** for reporting implementations.
- Error handling while sending reports instead of panicking.

## Linear Ticket

https://linear.app/rudderstack/issue/PIPE-316/create-a-new-reporting-implementation-that-will-filter-transform-and

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: enterprise/reporting/error_index_reporting.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	"github.com/rudderlabs/rudder-server/utils/types"
+)
+
+const (
+	ei = "ei"
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:21]: Please find a short name, easy to remember and pronounce :)

File: enterprise/reporting/error_index_reporting.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+				MessageID:        failedMessage.MessageID,
+				SourceID:         metric.SourceID,
+				DestinationID:    metric.DestinationID,
+				TransformationID: metric.TransformationID,
+				TrackingPlanID:   metric.TrackingPlanID,
+				FailedStage:      metric.PUDetails.PU,
+				EventName:        metric.StatusDetail.EventName,
+				EventType:        metric.StatusDetail.EventType,
+				ReceivedAt:       failedMessage.ReceivedAt,
+				FailedAt:         eir.now(),
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:107]: **minor:** let's use the same time for all messages in the batch, i.e. calculate `failedAt := eir.now()` before starting the loop

File: enterprise/reporting/error_index_reporting.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+			}
+			payloadJSON, err := json.Marshal(payload)
+			if err != nil {
+				panic(fmt.Errorf("unable to marshal payload: %v", err))
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:111]: errors should be propagated

File: enterprise/reporting/error_index_reporting.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			}
+
+			params := struct {
+				WorkspaceID string `json:"workspaceId"`
+				SourceID    string `json:"sourceId"`
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:116]:
```suggestion
				SourceID    string `json:"source_id"`
```

File: enterprise/reporting/error_index_reporting.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+			}
+
+			jobs = append(jobs, &jobsdb.JobT{
+				UUID:         uuid.New(),
+				UserID:       uuid.New().String(),
+				Parameters:   paramsJSON,
+				CustomVal:    ei,
+				EventPayload: payloadJSON,
+				EventCount:   0,
+				WorkspaceId:  eir.configSubscriber.WorkspaceIDFromSource(metric.SourceID),
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:133]:
```suggestion
				WorkspaceId:  workspaceID,
```

File: enterprise/reporting/error_index_reporting.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+				panic(fmt.Errorf("unable to marshal params: %v", err))
+			}
+
+			jobs = append(jobs, &jobsdb.JobT{
+				UUID:         uuid.New(),
+				UserID:       uuid.New().String(),
+				Parameters:   paramsJSON,
+				CustomVal:    ei,
+				EventPayload: payloadJSON,
+				EventCount:   0,
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:132]:
```suggestion
				EventCount:   1,
```

File: enterprise/reporting/error_index_reporting.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+				SourceID:    metric.SourceID,
+			}
+			paramsJSON, err := json.Marshal(params)
+			if err != nil {
+				panic(fmt.Errorf("unable to marshal params: %v", err))
+			}
+
+			jobs = append(jobs, &jobsdb.JobT{
+				UUID:         uuid.New(),
+				UserID:       uuid.New().String(),
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:128]: Not important, can use an empty string
```suggestion
```

File: enterprise/reporting/error_index_reporting.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+				Parameters:   paramsJSON,
+				CustomVal:    ei,
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:130]: Not important, can use an empty string

```suggestion
```

File: enterprise/reporting/error_index_reporting.go (hunk 1 above)
Comment (achettyiitr) [reply enterprise/reporting/error_index_reporting.go]: Does `eridx` feels ok?

Comment (atzoum) [reply enterprise/reporting/error_index_reporting.go]: `err_idx`?

**Review by atzoum — APPROVED** [review]

File: enterprise/reporting/error_index_reporting.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+	SourceID         string    `json:"sourceId"`
+	DestinationID    string    `json:"destinationId"`
+	TransformationID string    `json:"transformationId"`
+	TrackingPlanID   string    `json:"trackingPlanId"`
+	FailedStage      string    `json:"failedStage"`
+	EventType        string    `json:"eventType"`
+	EventName        string    `json:"eventName"`
+	ReceivedAt       time.Time `json:"receivedAt"`
+	FailedAt         time.Time `json:"failedAt"`
+}
```
Comment (atzoum) [inline enterprise/reporting/error_index_reporting.go:34]: @cisse21 do these cover our own current needs?

File: processor/processor.go [hunk 10]
```diff
@@ -2126,7 +2126,10 @@ func (proc *Handle) Store(partition string, in *storeMessage) {
 			}
 
 			if proc.isReportingEnabled() {
-				proc.reporting.Report(in.reportMetrics, tx.SqlTx())
+				err = proc.reporting.Report(in.reportMetrics, tx.SqlTx())
+				if err != nil {
+					return fmt.Errorf("reporting metrics: %w", err)
+				}
```
Comment (Sidddddarth) [inline processor/processor.go:2132]: minor:
```suggestion
				if err = proc.reporting.Report(in.reportMetrics, tx.SqlTx()) {
					return fmt.Errorf("reporting metrics: %w", err)
				}
```

File: router/batchrouter/handle.go [hunk 11]
```diff
@@ -755,7 +755,10 @@ func (brt *Handle) updateJobStatus(batchJobs *BatchedJobs, isWarehouse bool, err
 			}
 
 			if brt.reporting != nil && brt.reportingEnabled {
-				brt.reporting.Report(reportMetrics, tx.SqlTx())
+				err = brt.reporting.Report(reportMetrics, tx.SqlTx())
+				if err != nil {
+					return fmt.Errorf("reporting metrics: %w", err)
+				}
```
Comment (Sidddddarth) [inline router/batchrouter/handle.go:761]: minor
```suggestion
				if err = brt.reporting.Report(reportMetrics, tx.SqlTx()); err != nil {
					return fmt.Errorf("reporting metrics: %w", err)
				}
```

File: enterprise/reporting/mediator.go [hunk 12]
```diff
@@ -12,16 +12,16 @@ import (
 	"github.com/rudderlabs/rudder-server/utils/types"
 )
 
-type ReportingMediator struct {
+type Mediator struct {
 	log logger.Logger
 
 	g         *errgroup.Group
 	ctx       context.Context
 	reporters []types.Reporting
```
Comment (Sidddddarth) [inline enterprise/reporting/mediator.go:20]: minor preference
```suggestion
	reporters []types.Reporter
```

Comment (achettyiitr) [reply enterprise/reporting/mediator.go:20]: Is it ok if I handle this in a separate PR?

File: enterprise/reporting/error_index_reporting.go (hunk 9 above)
Comment (cisse21) [reply enterprise/reporting/error_index_reporting.go:34]: Yeah this looks good for all the current needs

**Review by Sidddddarth — APPROVED** [review]

**Pull request merged**